hpx/traits/action_put_parcel.hpp

```diff
@@ -0,0 +1,149 @@
+//  Copyright (c) 2016 John Biddiscombe
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_TRAITS_ACTION_PUT_PARCEL_AUG_2016)
+#define HPX_TRAITS_ACTION_PUT_PARCEL_AUG_2016
+
+#include <hpx/config.hpp>
+#include <hpx/runtime/threads/thread_enums.hpp>
+#include <hpx/runtime/naming/id_type.hpp>
+#include <hpx/runtime/naming/address.hpp>
+
+namespace hpx { namespace traits
+{
+    namespace detail {
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename Action>
+        inline naming::address&& complement_addr(naming::address& addr)
+        {
+            if (components::component_invalid == addr.type_)
+            {
+                addr.type_ = components::get_component_type<
```

Comment:
The setting of component type should be done in the put_parcel functions already.